sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java

```diff
@@ -0,0 +1,457 @@
+  }
+
+  private static Class convertPrimitiveToSqlType(RelDataType type)
+  {
+    /* I hope that following this method instead of calling value.value() is better
+    because we can catch any type mismatches. */
+    switch (type.getSqlTypeName()) {
+      case BOOLEAN:
+        return Boolean.class;
+      case TINYINT:
+      case SMALLINT:
+      case INTEGER:
+        return Integer.class;
+      case BIGINT:
+        return Long.class;
+      case REAL:
+        return Float.class;
+      case FLOAT:
+      case DOUBLE:
+        return Double.class;
+      case DATE:
+        return Date.class;
+      case TIME:
+        return Date.class;
+      case TIMESTAMP:
+        return Date.class;
+      case CHAR:
+      case VARCHAR:
+        return String.class;
+      case BINARY:
+      case VARBINARY:
+        return Byte.class;
+      case ANY:
+      case SYMBOL:
+        return Object.class;
+      default:
+        throw new RuntimeException(String.format("Unsupported type %s", type.getSqlTypeName()));
+    }
+  }
+
+  public final RelInfo traverse(RelNode n) throws Exception
+  {
+    List<RelInfo> inputStreams = new ArrayList<>();
+    for (RelNode input : n.getInputs()) {
+      inputStreams.add(traverse(input));
+    }
+
+    RelInfo currentNodeRelInfo;
+
+    if (n instanceof Aggregate) {
+      currentNodeRelInfo = visitAggregate((Aggregate)n, inputStreams);
+    } else if (n instanceof Calc) {
+      currentNodeRelInfo = visitCalc((Calc)n, inputStreams);
+    } else if (n instanceof Collect) {
+      currentNodeRelInfo = visitCollect((Collect)n, inputStreams);
+    } else if (n instanceof Correlate) {
+      currentNodeRelInfo = visitCorrelate((Correlate)n, inputStreams);
+    } else if (n instanceof Delta) {
+      currentNodeRelInfo = visitDelta((Delta)n, inputStreams);
+    } else if (n instanceof Exchange) {
+      currentNodeRelInfo = visitExchange((Exchange)n, inputStreams);
+    } else if (n instanceof Project) {
+      currentNodeRelInfo = visitProject((Project)n, inputStreams);
+    } else if (n instanceof Filter) {
+      currentNodeRelInfo = visitFilter((Filter)n, inputStreams);
+    } else if (n instanceof Sample) {
+      currentNodeRelInfo = visitSample((Sample)n, inputStreams);
+    } else if (n instanceof Sort) {
+      currentNodeRelInfo = visitSort((Sort)n, inputStreams);
+    } else if (n instanceof TableModify) {
+      currentNodeRelInfo = visitTableModify((TableModify)n, inputStreams);
+    } else if (n instanceof TableScan) {
+      currentNodeRelInfo = visitTableScan((TableScan)n, inputStreams);
+    } else if (n instanceof Uncollect) {
+      currentNodeRelInfo = visitUncollect((Uncollect)n, inputStreams);
+    } else if (n instanceof Window) {
+      currentNodeRelInfo = visitWindow((Window)n, inputStreams);
+    } else if (n instanceof Join) {
+      currentNodeRelInfo = visitJoin((Join)n, inputStreams);
+    } else {
+      currentNodeRelInfo = defaultValue(n, inputStreams);
+    }
```

Comment:
Not sure why we should do that, this is post-traverse processing block which appears correct to be in this method.